subsys/nrf_security/src/ssf_secdom/sdfw_psa_ns_api.c

```diff
@@ -0,0 +1,141 @@
+
+psa_status_t psa_call_buffered_and_flushed(psa_handle_t handle, int32_t type,
+					   const psa_invec *in_vec, size_t in_len,
+					   psa_outvec *out_vec, size_t out_len)
+{
+	/* We have no need for this at this time */
+	ARG_UNUSED(type);
+
+	psa_status_t ipc_status = sdfw_psa_ns_ipc_setup();
+
+	if (ipc_status != PSA_SUCCESS) {
+		return ipc_status;
+	}
+
+	/* volatile and flushed because the cpusec core will usually
+	 * modify this variable
+	 */
+	psa_status_t volatile status = PSA_ERROR_COMMUNICATION_FAILURE;
+	sys_cache_data_flush_range((void *)&status, sizeof(status));
+
+	uint32_t ipc_service_buf[SDFW_PSA_IPC_DATA_LEN];
+
+	ipc_service_buf[0] = handle; /* i.e. TFM_CRYPTO_HANDLE defined to 0x40000100U */
+	ipc_service_buf[1] = (uint32_t)in_vec;
+	ipc_service_buf[2] = in_len;
+	ipc_service_buf[3] = (uint32_t)out_vec;
+	ipc_service_buf[4] = out_len;
+	ipc_service_buf[5] = (uint32_t)&status;
+
+	int32_t ret = sdfw_psa_ns_ipc_send(ipc_service_buf, sizeof(ipc_service_buf));
+
+	if (ret != sizeof(ipc_service_buf)) {
+		return PSA_ERROR_COMMUNICATION_FAILURE;
+	}
+
+	do {
+		sys_cache_data_flush_and_invd_range((void *)&status, sizeof(status));
+	} while (status == PSA_ERROR_COMMUNICATION_FAILURE);
+
+	return status;
+}
+
+/*
+ * Addresses that the cpusec reads and/or writes must be flushed
+ * before the PSA call.
+ *
+ * After the PSA call, any address that the cpusec may have written to
+ * must be flushed and invalidated.
+ */
+psa_status_t psa_call_buffered(psa_handle_t handle, int32_t type, const psa_invec *in_vec,
+			       size_t in_len, psa_outvec *out_vec, size_t out_len)
+{
+	for (int i = 0; i < out_len; i++) {
+		sys_cache_data_flush_range((void *)out_vec[i].base, out_vec[i].len);
+	}
+
+	for (int i = 0; i < in_len; i++) {
+		sys_cache_data_flush_range((void *)in_vec[i].base, in_vec[i].len);
+	}
+
+	sys_cache_data_flush_range((void *)in_vec, in_len * sizeof(in_vec[0]));
+	sys_cache_data_flush_range((void *)out_vec, out_len * sizeof(out_vec[0]));
+
+	psa_status_t status =
+		psa_call_buffered_and_flushed(handle, type, in_vec, in_len, out_vec, out_len);
+
+	for (int i = 0; i < out_len; i++) {
+		sys_cache_data_flush_and_invd_range(out_vec[i].base, out_vec[i].len);
+	}
+
+	/* cpusec may write the number of bytes writen to out_vec[i].len */
+	sys_cache_data_flush_and_invd_range((void *)out_vec, out_len * sizeof(out_vec[0]));
+
+	return status;
+}
+
+/*
+ * Both the start address and end address of buffers that the cpusec
+ * writes must be 4-byte aligned. bounce_buffers, when enabled with
+ * PSA_SSF_CRYPTO_CLIENT_OUT_BOUNCE_BUFFERS, will correct the
+ * alignment by buffering such buffers when necessary.
+ */
+psa_status_t psa_call(psa_handle_t handle, int32_t type, const psa_invec *in_vec, size_t in_len,
+		      psa_outvec *out_vec, size_t out_len)
+{
+	if (in_len > PSA_MAX_IOVEC || out_len > PSA_MAX_IOVEC) {
+		return PSA_ERROR_PROGRAMMER_ERROR;
+	}
+
+	psa_status_t status;
+
+	psa_outvec out_vec_copy[out_len];
```

Comment:
how does this work? doesn't the declaration need to have a constant size at build time?

Reply:
I was a bit worried the compiler would put in a malloc here.

But I see that it modifies the stack pointer to have the stack usage be dynamic based on the size of out_len. (See the mov instruction modifying sp (stack pointer)).

In the general case dynamic stack size is scary. But in this case out_len is verified to be small, and sizeof(psa_outvec) is  small so we are only dynamically increasing the stack size a little bit.

![image](https://github.com/user-attachments/assets/05a19463-3fc7-4243-a331-b38289193054)
https://gcc.gnu.org/onlinedocs/gcc/Variable-Length.html

Reply:
huh neat, good to know!